<a href="https://colab.research.google.com/github/LucasA0911/Analise_de_dados/blob/main/aula_modulo4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 1. Começando com a biblioteca Pandas

## Importação da biblioteca

In [ ]:
# Importanto a biblioteca
import pandas as pd

## Gerando dados para manipular com o pandas

In [ ]:
dados = {
    'coluna-1': [1,2,3,4,5],
    'coluna-2': ['a','b','c','d','e'],
    'coluna-3': [1.1,2.2,3.3,4.4,5.5]
}

In [ ]:
dados

{'coluna-1': [1, 2, 3, 4, 5],
 'coluna-2': ['a', 'b', 'c', 'd', 'e'],
 'coluna-3': [1.1, 2.2, 3.3, 4.4, 5.5]}

In [ ]:
df = pd.DataFrame(dados)

In [ ]:
df

,coluna-1,coluna-2,coluna-3
0,1,a,1.1
1,2,b,2.2
2,3,c,3.3
3,4,d,4.4
4,5,e,5.5


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   coluna-1  5 non-null      int64  
 1   coluna-2  5 non-null      object 
 2   coluna-3  5 non-null      float64
dtypes: float64(1), int64(1), object(1)
memory usage: 252.0+ bytes


# 2. Começando com a biblioteca Numpy

## Importação da biblioteca

In [ ]:
import numpy as np

## Criação de um array de 2 dimensões

In [ ]:
dadosnp = np.array([[1,2,3],[4,5,6]])

In [ ]:
dadosnp

array([[1, 2, 3],
       [4, 5, 6]])

In [ ]:
dadosnp2 = np.arange(15, dtype=np.int64).reshape(3,5)

In [ ]:
dadosnp2

array([[ 0,  1,  2,  3,  4],
       [ 5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14]])

# Praticando com o pandas

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/books.csv', sep=',',on_bad_lines='skip')

In [ ]:
df.head()

,bookID,title,authors,average_rating,isbn,isbn13,language_code,num_pages,ratings_count,text_reviews_count,publication_date,publisher
0,1,Harry Potter and the Half-Blood Prince (Harry ...,J.K. Rowling/Mary GrandPré,4.57,0439785960,9780439785969,eng,652,2095690,27591,9/16/2006,Scholastic Inc.
1,2,Harry Potter and the Order of the Phoenix (Har...,J.K. Rowling/Mary GrandPré,4.49,0439358078,9780439358071,eng,870,2153167,29221,9/1/2004,Scholastic Inc.
2,4,Harry Potter and the Chamber of Secrets (Harry...,J.K. Rowling,4.42,0439554896,9780439554893,eng,352,6333,244,11/1/2003,Scholastic
3,5,Harry Potter and the Prisoner of Azkaban (Harr...,J.K. Rowling/Mary GrandPré,4.56,043965548X,9780439655484,eng,435,2339585,36325,5/1/2004,Scholastic Inc.
4,8,Harry Potter Boxed Set Books 1-5 (Harry Potte...,J.K. Rowling/Mary GrandPré,4.78,0439682584,9780439682589,eng,2690,41428,164,9/13/2004,Scholastic


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11123 entries, 0 to 11122
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   bookID              11123 non-null  int64  
 1   title               11123 non-null  object 
 2   authors             11123 non-null  object 
 3   average_rating      11123 non-null  float64
 4   isbn                11123 non-null  object 
 5   isbn13              11123 non-null  int64  
 6   language_code       11123 non-null  object 
 7     num_pages         11123 non-null  int64  
 8   ratings_count       11123 non-null  int64  
 9   text_reviews_count  11123 non-null  int64  
 10  publication_date    11123 non-null  object 
 11  publisher           11123 non-null  object 
dtypes: float64(1), int64(5), object(6)
memory usage: 1.0+ MB


In [ ]:
# Verificando linhas com amostras aleatórias

df.sample(5)

,bookID,title,authors,average_rating,isbn,isbn13,language_code,num_pages,ratings_count,text_reviews_count,publication_date,publisher
5327,19282,The Odyssey,Homer/Robert Fitzgerald/Seamus Heaney,3.76,1857150945,9781857150940,eng,509,48,3,10/8/1992,Everyman
2839,10559,The Three Musketeers,Alexandre Dumas/Barbara T. Cooper,4.07,1593081480,9781593081485,eng,761,1149,147,11/25/2004,Barnes Noble Classics
9790,39286,Hide-and-Seek with Angels: The Life of J.M. Ba...,Lisa Chaney,3.91,0099453231,9780099453239,eng,448,66,11,5/16/2006,Arrow
2674,9828,Rat Bastards: The Life and Times of South Bost...,"John ""Red"" Shea/Mark Wahlberg",3.51,0060837160,9780060837167,eng,304,34,3,3/14/2006,William Morrow
11110,45607,Las Crónicas de Narnia,C.S. Lewis/Margarita E. Valdes/Gemma Gallart/P...,4.26,0061199001,9780061199004,spa,816,186,11,11/7/2006,HarperCollins Espanol


In [44]:
# Linhas finais do dataset
df.tail()

,bookID,title,authors,average_rating,isbn,isbn13,language_code,num_pages,ratings_count,text_reviews_count,publication_date,publisher
11118,45631,Expelled from Eden: A William T. Vollmann Reader,William T. Vollmann/Larry McCaffery/Michael He...,4.06,1560254416,9781560254416,eng,512,156,20,12/21/2004,Da Capo Press
11119,45633,You Bright and Risen Angels,William T. Vollmann,4.08,0140110879,9780140110876,eng,635,783,56,12/1/1988,Penguin Books
11120,45634,The Ice-Shirt (Seven Dreams #1),William T. Vollmann,3.96,0140131965,9780140131963,eng,415,820,95,8/1/1993,Penguin Books
11121,45639,Poor People,William T. Vollmann,3.72,0060878827,9780060878825,eng,434,769,139,2/27/2007,Ecco
11122,45641,Las aventuras de Tom Sawyer,Mark Twain,3.91,8497646983,9788497646987,spa,272,113,12,5/28/2006,Edimat Libros


In [ ]:
# verifica se tem dados nulos
df.isnull().sum()

,0
bookID,0
title,0
authors,0
average_rating,0
isbn,0
isbn13,0
language_code,0
num_pages,0
ratings_count,0
text_reviews_count,0


## Escolhendo as caracteristicas relevantes para a análise

In [25]:
df_filtrado = df[['title','authors','average_rating']]
df_filtrado.head()

,title,authors,average_rating
0,Harry Potter and the Half-Blood Prince (Harry ...,J.K. Rowling/Mary GrandPré,4.57
1,Harry Potter and the Order of the Phoenix (Har...,J.K. Rowling/Mary GrandPré,4.49
2,Harry Potter and the Chamber of Secrets (Harry...,J.K. Rowling,4.42
3,Harry Potter and the Prisoner of Azkaban (Harr...,J.K. Rowling/Mary GrandPré,4.56
4,Harry Potter Boxed Set Books 1-5 (Harry Potte...,J.K. Rowling/Mary GrandPré,4.78


In [26]:
maior_classificado = df['average_rating'].max()
menor_classificado = df['average_rating'].min()
media_classificado = df['average_rating'].mean()
moda_classificado = df['average_rating'].mode()
mediana_classificado = df['average_rating'].median()


In [29]:
# Maior avaliação
maior_classificado

5.0

In [30]:
# Menor avaliação
menor_classificado

0.0

In [37]:
# Media das avaliações
media_classificado.round(2)

np.float64(3.93)

In [32]:
# Moda das avaliações
moda_classificado

,average_rating
0,4.0


In [33]:
# Mediana das avaliações
mediana_classificado

3.96

In [38]:
# Calculando a variância
df['average_rating'].var()

0.12283995368329198

In [40]:
# Calculando o desvio padrão
df['average_rating'].std()

0.3504853116512759

In [27]:
df.describe()

,bookID,average_rating,isbn13,num_pages,ratings_count,text_reviews_count
count,11123.000000,11123.000000,1.112300e+04,11123.000000,1.112300e+04,11123.000000
mean,21310.856963,3.934075,9.759880e+12,336.405556,1.794285e+04,542.048099
std,13094.727252,0.350485,4.429758e+11,241.152626,1.124992e+05,2576.619589
min,1.000000,0.000000,8.987060e+09,0.000000,0.000000e+00,0.000000
25%,10277.500000,3.770000,9.780345e+12,192.000000,1.040000e+02,9.000000
50%,20287.000000,3.960000,9.780582e+12,299.000000,7.450000e+02,47.000000
75%,32104.500000,4.140000,9.780872e+12,416.000000,5.000500e+03,238.000000
max,45641.000000,5.000000,9.790008e+12,6576.000000,4.597666e+06,94265.000000


In [42]:
# livros que possuem mais comentários do que o número de reviews
df[df.ratings_count < df.text_reviews_count]

,bookID,title,authors,average_rating,isbn,isbn13,language_code,num_pages,ratings_count,text_reviews_count,publication_date,publisher
7209,27647,Secretos De Familia,Julia Glass,3.57,8496525031,9788496525030,spa,528,0,1,11/1/2006,Puzzle-Roca
10138,41044,Day and Night,Better Homes and Gardens,0.00,0696018829,9780696018824,eng,32,0,1,3/1/1989,Meredith Corporation


In [43]:
# livros que possuem  mais número de reviews do que o número de comentários
df[df.ratings_count > df.text_reviews_count]

,bookID,title,authors,average_rating,isbn,isbn13,language_code,num_pages,ratings_count,text_reviews_count,publication_date,publisher
0,1,Harry Potter and the Half-Blood Prince (Harry ...,J.K. Rowling/Mary GrandPré,4.57,0439785960,9780439785969,eng,652,2095690,27591,9/16/2006,Scholastic Inc.
1,2,Harry Potter and the Order of the Phoenix (Har...,J.K. Rowling/Mary GrandPré,4.49,0439358078,9780439358071,eng,870,2153167,29221,9/1/2004,Scholastic Inc.
2,4,Harry Potter and the Chamber of Secrets (Harry...,J.K. Rowling,4.42,0439554896,9780439554893,eng,352,6333,244,11/1/2003,Scholastic
3,5,Harry Potter and the Prisoner of Azkaban (Harr...,J.K. Rowling/Mary GrandPré,4.56,043965548X,9780439655484,eng,435,2339585,36325,5/1/2004,Scholastic Inc.
4,8,Harry Potter Boxed Set Books 1-5 (Harry Potte...,J.K. Rowling/Mary GrandPré,4.78,0439682584,9780439682589,eng,2690,41428,164,9/13/2004,Scholastic
...,...,...,...,...,...,...,...,...,...,...,...,...
11118,45631,Expelled from Eden: A William T. Vollmann Reader,William T. Vollmann/Larry McCaffery/Michael He...,4.06,1560254416,9781560254416,eng,512,156,20,12/21/2004,Da Capo Press
11119,45633,You Bright and Risen Angels,William T. Vollmann,4.08,0140110879,9780140110876,eng,635,783,56,12/1/1988,Penguin Books
11120,45634,The Ice-Shirt (Seven Dreams #1),William T. Vollmann,3.96,0140131965,9780140131963,eng,415,820,95,8/1/1993,Penguin Books
11121,45639,Poor People,William T. Vollmann,3.72,0060878827,9780060878825,eng,434,769,139,2/27/2007,Ecco


In [46]:
# Autores mais influentes
autores = df.groupby('authors').agg({'average_rating':'mean'}).sort_values(by='average_rating',ascending=False)
autores.head(10)

,average_rating
authors,
R. McL. Wilson,5.0
Elena N. Mahlow,5.0
John Diamond,5.0
Julie Sylvester/David Sylvester,5.0
William C. Dowling,5.0
James E. Campbell,5.0
Todd Davis/Marc Frey,5.0
Middlesex Borough Heritage Committee,5.0
Dennis Adler/R.L. Wilson,5.0


In [47]:
# Autores mais influentes agrupados por autor, quantidade de livros e média de avaliação

autores = df.groupby('authors').agg(
    total_livros = ('title','count'),
    media_avaliacao = ('average_rating','mean')
).sort_values(by='total_livros', ascending= False)
autores.head(10)


,total_livros,media_avaliacao
authors,,
P.G. Wodehouse,40,4.147250
Stephen King,40,3.973500
Rumiko Takahashi,39,4.186667
Orson Scott Card,35,3.779143
Agatha Christie,33,3.989091
Piers Anthony,30,3.716667
Sandra Brown,29,3.742069
Mercedes Lackey,29,4.039310
Dick Francis,28,3.962143


In [48]:
# Analise por idioma:
idiomas = df.groupby('language_code').agg(
    total_livros = ('title','count'),
    media_avaliacao = ('average_rating','mean')
).sort_values(by='total_livros', ascending= False)
idiomas.head(10)

,total_livros,media_avaliacao
language_code,,
eng,8908,3.934062
en-US,1408,3.914659
spa,218,3.929312
en-GB,214,3.923411
fre,144,3.971528
ger,99,3.950101
jpn,46,4.268696
mul,19,4.126316
zho,14,4.456429


# Relatório:

O conjunto de dados `books.csv` contém informações sobre uma variedade de livros. As colunas incluem `bookID`, `title`, `authors`, `average_rating`, `isbn`, `isbn13`, `language_code`, `num_pages`, `ratings_count`, `text_reviews_count`, `publication_date` e `publisher`.

## Análise dos Dados

### Carregamento e Inspeção Inicial

O dataset foi carregado a partir do arquivo `books.csv`. Uma inspeção inicial (`df.info()`) revelou que o dataset possui **11.123 entradas** e **12 colunas**. Todas as colunas possuem valores não nulos, indicando que não há dados faltantes. Os tipos de dados incluem inteiros (`int64`), objetos (`object`) e ponto flutuante (`float64`).

### Estatísticas Descritivas

Abaixo estão as estatísticas descritivas para as colunas numéricas do dataset, como `bookID`, `average_rating`, `isbn13`, `num_pages`, `ratings_count` e `text_reviews_count`.

In [49]:
display(df.describe())

,bookID,average_rating,isbn13,num_pages,ratings_count,text_reviews_count
count,11123.000000,11123.000000,1.112300e+04,11123.000000,1.112300e+04,11123.000000
mean,21310.856963,3.934075,9.759880e+12,336.405556,1.794285e+04,542.048099
std,13094.727252,0.350485,4.429758e+11,241.152626,1.124992e+05,2576.619589
min,1.000000,0.000000,8.987060e+09,0.000000,0.000000e+00,0.000000
25%,10277.500000,3.770000,9.780345e+12,192.000000,1.040000e+02,9.000000
50%,20287.000000,3.960000,9.780582e+12,299.000000,7.450000e+02,47.000000
75%,32104.500000,4.140000,9.780872e+12,416.000000,5.000500e+03,238.000000
max,45641.000000,5.000000,9.790008e+12,6576.000000,4.597666e+06,94265.000000


Principais observações das estatísticas descritivas:

*   **`average_rating`**: A classificação média varia de 0.0 a 5.0, com uma média de aproximadamente **3.93**. A moda (valor mais frequente) é **4.0** e a mediana é **3.96**. A variância é de **0.12** e o desvio padrão é de **0.35**, indicando que as classificações não estão amplamente dispersas.
*   **`num_pages`**: O número de páginas varia bastante, de 0 a 6576, com uma média de cerca de **336 páginas**.
*   **`ratings_count`**: O número de avaliações é muito variável, com alguns livros tendo zero avaliações e outros com milhões, indicando grande popularidade de certos títulos. A média é de aproximadamente 17.942 avaliações.
*   **`text_reviews_count`**: Similarmente, a contagem de resenhas de texto também varia significativamente, com uma média de 542 resenhas.

### Análise de Autores

Foi realizada uma análise para identificar os autores mais influentes, considerando o número total de livros e a média de suas avaliações. Os 10 autores com mais livros no dataset são:

In [50]:
display(autores.head(10))

,total_livros,media_avaliacao
authors,,
P.G. Wodehouse,40,4.147250
Stephen King,40,3.973500
Rumiko Takahashi,39,4.186667
Orson Scott Card,35,3.779143
Agatha Christie,33,3.989091
Piers Anthony,30,3.716667
Sandra Brown,29,3.742069
Mercedes Lackey,29,4.039310
Dick Francis,28,3.962143


Podemos observar que **P.G. Wodehouse** e **Stephen King** lideram em número de livros (40 cada), com classificações médias respeitáveis.

### Análise por Idioma

A distribuição de livros e suas classificações médias também foi investigada por `language_code`. Os 10 idiomas com mais livros no dataset são:

In [51]:
display(idiomas.head(10))

,total_livros,media_avaliacao
language_code,,
eng,8908,3.934062
en-US,1408,3.914659
spa,218,3.929312
en-GB,214,3.923411
fre,144,3.971528
ger,99,3.950101
jpn,46,4.268696
mul,19,4.126316
zho,14,4.456429


O idioma **inglês (`eng`)** é predominante, representando a vasta maioria dos livros, seguido por `en-US` e `spa` (espanhol). As classificações médias são consistentemente altas entre os idiomas mais comuns.

### Discrepâncias em Avaliações e Comentários

Foi identificado um pequeno número de livros onde o `ratings_count` (contagem de avaliações) é menor que o `text_reviews_count` (contagem de resenhas de texto). Isso pode indicar anomalias nos dados ou casos específicos onde resenhas foram registradas sem uma avaliação numérica correspondente. Por outro lado, a vasta maioria dos livros segue a lógica esperada, com mais avaliações do que resenhas.

## Conclusão

O dataset `books` é um recurso rico para análise, com informações detalhadas sobre livros, autores e avaliações. As análises realizadas fornecem uma visão geral da distribuição de classificações, a proeminência de certos autores e a diversidade linguística dos livros. Futuras análises poderiam explorar a correlação entre o número de páginas e a classificação, a evolução das classificações ao longo do tempo ou a influência de editores específicos.